### PR TITLE
Adds a stationloving component to replace a couple of flags and SSinbounds with

### PR DESCRIPTION
**Pull request by vuonojenmustaturska**
:cl: Naksu
code: Inbounds subsystem, STATIONLOVING_2 and INFORM_ADMINS_ON_RELOCATE_2 flags are removed.
code: New stationloving component replaces all of this.
code: Added new signals: COMSIG_MOVABLE_Z_CHANGED is now sent when a movable changes Z level, COMSIG_PARENT_PREQDELETED is sent before the object's Destroy() is called and allows interrupting the deletion, COMSIG_ITEM_IMBUE_SOUL is sent when a wizard attempts to make a thing their phylactery
/:cl:

Not packing the flags to erase the gap I left in there because I intend to take an axe to several of them.

image
tested in-game with both destroys and moves

### CODE REVIEW

File: code/datums/components/stationloving.dm [hunk 1]
```diff
@@ -0,0 +1,69 @@
...
+	src.inform_admins = inform_admins
+	check_in_bounds() // Just in case something is being created outside of station/centcom
+
+/datum/component/stationloving/proc/relocate()
+	var/targetturf = find_safe_turf()
+	if(!targetturf)
+		if(GLOB.blobstart.len > 0)
+			targetturf = get_turf(pick(GLOB.blobstart))
+		else
+			throw EXCEPTION("Unable to find a blobstart landmark")
```
Comment (Cyberboss) [inline code/datums/components/stationloving.dm]: CRASH

File: code/datums/components/stationloving.dm [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+		AM.forceMove(targetturf)
+	// move the disc, so ghosts remain orbiting it even if it's "destroyed"
+	return targetturf
+
+/datum/component/stationloving/proc/check_in_bounds()
+	if(in_bounds())
+		return
+	else
+		var/turf/currentturf = get_turf(src)
+		to_chat(get(parent, /mob), "<span class='danger'>You can't help but feel that you just lost something back there...</span>")
```
Comment (Cyberboss) [inline code/datums/components/stationloving.dm]: ?

File: code/game/atoms_movable.dm [hunk 3]
```diff
@@ -251,24 +251,12 @@
...
 
 /atom/movable/Destroy(force)
-	var/inform_admins = (flags_2 & INFORM_ADMINS_ON_RELOCATE_2)
-	var/stationloving = (flags_2 & STATIONLOVING_2)
 
-	if(inform_admins && force)
-		var/turf/T = get_turf(src)
-		message_admins("[src] has been !!force deleted!! in [ADMIN_COORDJMP(T)].")
-		log_game("[src] has been !!force deleted!! in [COORD(T)].")
+	var/datum/component/stationloving/component = GetComponent(/datum/component/stationloving)
```
Comment (Cyberboss) [inline code/game/atoms_movable.dm]: Use COMSIG_PARENT_QDELETED, might need to modify it a bit to allow it to change the qdel hint

**Review by Cyberboss — CHANGES_REQUESTED** [review]

File: code/datums/components/stationloving.dm (hunk 2 above)
Comment (vuonojenmustaturska) [reply code/datums/components/stationloving.dm]: `get(atom/loc, type)` returns a parent of the atom `loc` that is of type `type`, this is the message that informs you you had a stationloving thing in your inventory when you went to lavaland or across a Z transition or whatever

File: code/modules/spells/spell_types/lichdom.dm [hunk 4]
```diff
@@ -34,7 +34,7 @@
 		for(var/obj/item in hand_items)
 			// I ensouled the nuke disk once. But it's probably a really
 			// mean tactic, so probably should discourage it.
-			if((item.flags_1 & ABSTRACT_1) || (item.flags_1 & NODROP_1) || (item.flags_2 & STATIONLOVING_2))
+			if((item.flags_1 & ABSTRACT_1) || (item.flags_1 & NODROP_1) || item.GetComponent(/datum/component/stationloving))
```
Comment (ninjanomnom) [inline code/modules/spells/spell_types/lichdom.dm]: Send a signal like `COMSIG_ITEM_IMBUE_SOUL` and make a return value able to cancel it

**Review by ninjanomnom — CHANGES_REQUESTED** [review]

File: code/game/atoms_movable.dm [hunk 5]
```diff
@@ -655,64 +638,17 @@
...
-	return targetturf
+	var/datum/component/stationloving/component = GetComponent(/datum/component/stationloving)
 
-/atom/movable/proc/check_in_bounds()
-	if(in_bounds())
+	if(state)
+		if (!component)
+			AddComponent(/datum/component/stationloving, inform_admins)
+		else if (component.inform_admins != inform_admins)
+			AddComponent(/datum/component/stationloving, inform_admins)
```
Comment (ninjanomnom) [inline code/game/atoms_movable.dm]: You're replicating work done by AddComponent here, just remove this proc and make the component use `COMPONENT_DUPE_UNIQUE_PASSARGS` to detect when inform_admins has changed

**Review by ninjanomnom — CHANGES_REQUESTED** [review]

File: code/game/atoms_movable.dm [hunk 6]
```diff
@@ -655,64 +638,13 @@
 */
 
 /atom/movable/proc/set_stationloving(state, inform_admins=FALSE)
```
Comment (ninjanomnom) [inline code/game/atoms_movable.dm]: I feel like this should be removed now, if an admin wants to use this feature they can add the component themselves or remove it from the component list.

**Pull request merged**